// ngx-bootstrap-datetime-popup/dist/datetime-popup.module.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { DatepickerModule } from 'ngx-bootstrap/datepicker';
import { TimepickerModule } from 'ngx-bootstrap/timepicker';
import { BsDropdownModule } from 'ngx-bootstrap/dropdown';
import { DatetimePopupComponent, OffClickDirective } from './components';
import { DatetimePopupService } from './services/datetime-popup';
var DatetimePopupModule = /** @class */ (function () {
    function DatetimePopupModule() {
    }
    DatetimePopupModule_1 = DatetimePopupModule;
    DatetimePopupModule.forRoot = function () {
        return {
            ngModule: DatetimePopupModule_1,
            providers: [
                DatetimePopupService
            ]
        };
    };
    var DatetimePopupModule_1;
    DatetimePopupModule = DatetimePopupModule_1 = __decorate([
        NgModule({
            imports: [
                CommonModule,
                FormsModule,
                DatepickerModule,
                TimepickerModule,
                BsDropdownModule
            ],
            declarations: [
                DatetimePopupComponent,
                OffClickDirective
            ],
            exports: [
                DatetimePopupComponent
            ],
            entryComponents: [
                DatetimePopupComponent
            ]
        })
    ], DatetimePopupModule);
    return DatetimePopupModule;
}());
export { DatetimePopupModule };
